Cloud deploy scripts AWS

fklezin opened the pull request:
DIA project

fklezin committed:
First commit

fklezin committed:
Added scripts for deployment of SystemDS on the Amazon EMR

fklezin committed:
removed unnecessary comments

fklezin committed:
formatted Readme.md

fklezin committed:
formatted Readme.md

Baunsgaard commented on scripts/aws/Readme.md:
```diff
@@ -0,0 +1,47 @@
+7. Terminate the EMR cluster: ./terminate_systemds_cluster.sh
+    
+####Further work
+#####Finetune the memory 
```
This part does not look good on GitHub as comments.

If you write: (newline between and space after # ) it improves

```code
#### Further work

##### Finetune the memory
```
Link to look at the result:

https://github.com/fklezin/systemds/tree/cloud_deploy_scripts/scripts/aws

fklezin replied:
Done

Baunsgaard commented on scripts/aws/aws_ec2_table.csv:
```diff
@@ -0,0 +1,117 @@
+﻿Name,	APIName,Memory,vCPUs,OnEMR,EMRcost
```
Maybe have a link in the README.md to the costs webpage. or am I Missing that this file is used for the scripts?

https://aws.amazon.com/emr/pricing/

fklezin replied:
It's feature that you can use.
When you specify the instance type in the config file, you parse this file to get memory and cpu of this instance type.
I realized it does not make sense to support this at the moment, because you need to specify the resources well: (https://aws.amazon.com/blogs/big-data/best-practices-for-successfully-managing-memory-for-apache-spark-applications-on-amazon-emr/)

So currently we are using: "maximizeResourceAllocation": "true", which sets spark default config for resource allocation to max.

fklezin replied:
It's feature that you can use.
When you specify the instance type in the config file, you parse this file to get memory and cpu of this instance type.
I realized that it does not make sense to support this at the moment: (https://aws.amazon.com/blogs/big-data/best-practices-for-successfully-managing-memory-for-apache-spark-applications-on-amazon-emr/)

So currently I'm using: "maximizeResourceAllocation": "true", which sets spark default config for resource allocation to max.

Baunsgaard commented on scripts/aws/run_systemds_script.sh:
```diff
@@ -0,0 +1,32 @@
+#! /bin/bash
+# $1 Dml script name.
+# $2 Args
```
You can make your script fail if there is not the correct arguments.

Like so:

```bash
# If number of arguments are not equal to 2.
if [ "$#" -ne 2 ] ; then
    # Make appropriate error / Help message
    echo "Usage: "$0" <DML script path> <Arguments>"
    exit
fi
```

fklezin replied:
Thanks. Done.

Baunsgaard commented on scripts/aws/spinup_systemds_cluster.sh:
```diff
@@ -0,0 +1,87 @@
+
+#Create the cluster
+CLUSTER_INFO=$(aws emr create-cluster \
+ --applications Name=Ganglia Name=Spark Name=Zeppelin \
```
Are you sure we need Ganglia and Zeppelin?

Baunsgaard commented on scripts/aws/systemds_cluster.config:
```diff
@@ -0,0 +1,25 @@
+
+KEYPAIR_NAME="SystemDSkeynamex"
+REGION="us-east-1"
+EMR_VERSION="emr-5.28.0"
```
I would add a comment specifying program versions associated with this EMR release.

https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-release-5x.html#emr-5290-release

```bash
# emr-5.28.0 -> spark 2.4.4
```

also newest version is 5.29.0 (hopefully soon to be 6.0.0)

fklezin replied:
Anyone can specify region, version, .. 
This is just the default version currently

Baunsgaard commented on scripts/aws/systemds_cluster.config:
```diff
@@ -0,0 +1,25 @@
+
+
+# Readonly #
+CLUSTER_ID=j-1NZJQA41A3NB
```
Maybe leave CLUSTER_ID blank?

Baunsgaard commented on scripts/aws/systemds_cluster.config:
```diff
@@ -0,0 +1,25 @@
+REGION="us-east-1"
+EMR_VERSION="emr-5.28.0"
+
+INSTANCES_TYPE="c5.xlarge"
```
Standard instance type c5.xlarge?

why not their General Purpose m5? (few cents more expensive)

Also it could be considered to go a generation back using the c4, or m4.

fklezin replied:
I set m5.xlarge as a default instance type in config.
c5 was used for testing, because it's cheaper.

Baunsgaard commented on scripts/aws/systemds_cluster.config:
```diff
@@ -0,0 +1,25 @@
+
+INSTANCES_TYPE="c5.xlarge"
+MASTER_INSTANCES_COUNT=1
+CORE_INSTANCES_COUNT=50
```
Core Instances Count?
is this the number of non master nodes?
if so, decrease.

Baunsgaard commented on scripts/aws/Readme.md:
```diff
@@ -0,0 +1,47 @@
+
+1. Create aws account / use your existing aws account
+
+2. Install aws-cli on your system (https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-linux.html)
```
Move link to newline

maybe indent.

fklezin replied:
Done.

Baunsgaard reviewed (commented):
Nice Deploy script.

only minor comments

- Standard naming conventions make the read me name "README.md"
- Formatting markdown: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#lists
- License missing in top of files.
  - for the README.md look at  (remember to remove the word modifications): <https://raw.githubusercontent.com/tugraz-isds/systemds/master/README.md>
  - for the *.sh look at (again remove the word modifications): <https://github.com/tugraz-isds/systemds/blob/master/bin/systemds-spark.sh>

Also question: Did you look at getting spot instances through the API? it might be cool to add, since they would be cheaper.

fklezin committed:
Formatted README.md
Added licence headers.
Removed Zeppelin application from create cluster command.
Cleared up config.

fklezin committed:
Removed modifications from copyright

Baunsgaard commented:
Thank You for your work, I will merge it in!
